FIP-0045 updates removing opt-in migration (#455)
filecoin-project/FIPs: FIPS/fip-0045.md

---

TODO:
- Spec token receiver hook payload schema for allocation requests
- Spec token receiver hook payload for extending allocations with new datacap

> - Market method for fetching allocation/claim ids for deals (needed?)

Reviewer: What's the potential use case here?

```

An allocation specifies a range of terms for which the provider may commit the data, between some
minimum and maximum. 
An allocation's maximum term must be at least as large as its minimum term.
Due to the implementation of term enforcement (details below), clients should leave a large buffer
between the minimum and maximum term to make the allocation practical for a provider.

Network policy limits the maximum value for an allocation's expiration,
so that an allocation cannot occupy non-useful DataCap indefinitely.

**Parameters**
- `MinimumVerifiedAllocationTerm`: The minimum allowed value for `TermMinimum` is 6 months.

> - `MaximumVerifiedAllocationTerm`: The maximum allowed value for `TermMaximum` is 5 years (can be increased by a future FIP).

Reviewer: ```suggestion
- `MaximumVerifiedAllocationTerm`: The maximum allowed value for `TermMaximum` is 5 years (5 years is the current sector max lifetime, both may be increased by a future FIP).
```

> - `MaximumVerifiedAllocationExpiration`: The maximum difference between `Expiration` and the epoch at which an allocation is created is 60 days.

Reviewer 2: Why 60 days? Is this the current expiration term foe DataCap?

Author: I chose 60 days as a reasonable balance between flexibility for providers scheduling deal-making, and the value being effective at reducing squatting. I'm very open to adjusting this value – please make a comment in the discussion thread if you would propose an alternative.

There's no current expiration, but it's worth introducing here because the changes make it practically a bit easier to squat on datacap (possible today, but more involved).

Reviewer 2: I don't have any specific value in mind. I was trying to understand the reasoning for the 60 days. It sounds reasonable to me as well.

```

An allocation specifies a range of terms for which the provider may commit the data, between some
minimum and maximum. 
An allocation's maximum term must be at least as large as its minimum term.
Due to the implementation of term enforcement (details below), clients should leave a large buffer
between the minimum and maximum term to make the allocation practical for a provider.

> An allocation must expire, so that it cannot occupy non-useful DataCap indefinitely.

Reviewer: ```suggestion
An allocation must have an expiration, so that it cannot occupy non-useful DataCap indefinitely.
```

    Allocations: HAMT[Address]HAMT[AllocationID]Allocation
	
    // Sequence number for allocation IDs.
    // The value `0` is reserved for "no allocation"
    NextAllocationId: AllocationID // uint64
}
```

#### Creation of allocations
An allocation is created by a verified client by transferring DataCap tokens (see below)
to the verified registry. 
The `Allocation` metadata must be provided as transfer metadata.
The verified registry checks the same preconditions as the existing `UseBytes` method. 
The size of the allocation must match the number of tokens received.

#### Removal of expired allocations

> An allocation may be removed after its expiration epoch has passed (by anyone). 

Reviewer 2: Why are we allowing expired allocations to be removed "by anyone"? What is the expected behavior here? Should clients remove them or are we supposed to have an automated process doing it?

Author: There is no reason to restrict it to the client address specifically, so this gives flexibility to users. E.g. if a client is a multisig or DAO or other complex entity, they don't need to go to the cost and trouble of multisigning this inconsequential message – one of the parties/delegates/employees can just do it.

There's no "should" or automation. Clients can clean up when the gas savings associated with a smaller collection justify it.

Reviewer 2: Got it. This makes sense.

Allocations are also removed when claimed by a provider (see below).

#### Revocation of un-allocated data cap

> The existing `RemoveVerifiedClientDataCap` method is changed to invoke a 

Reviewer 2: In what situations is this method called?

fn ClaimAllocations(params: ClaimAllocationsParams) -> ClaimAllocationsResult
```

#### Removal of expired claims

> The record of a claim may be removed after the term maximum has elapsed.

Reviewer 2: Who can remove expired claims? Is this an automated process?

Reviewer: Iiuc, Anyone can remove the expired claims and it’s triggered by an onchain message.

Author: Similar to above, anyone can. In practise this will probably be some ancillary or operations account associated with a provider. There's no automation.

fn RemoveExpiredClaims(params: RemoveExpiredClaimsParams)
```

The maximum term for a claim may be increased by the client which originally made the allocation,

> up to the network policy maximum.

Reviewer: could you please specify whats the `network policy maximum` referred here?

// Called by allocation client to extend the maximum term of their allocations, 
// without consuming DataCap.
// The new TermMaximum must be no less than the current one,
// and no greater than MaximumVerifiedAllocationTerm.
fn ExtendClaimTerms(params: ExtendClaimTermsParams) -> ExtendClaimTermsResult
```

#### Renewal of claim term by spending datacap
A verified client can extend the term for a claim _beyond_ the initial maximum term by
spending new DataCap.
The claim's term maximum can be extended up to `MaximumVerifiedAllocationTerm` beyond the current epoch.
The client extending the claim need not be the one that made the original allocation.
This is similar to issuing a new allocation/claim for the same data,

> but avoids overheads of re-sealing.

Reviewer 2: This mechanism makes a lot of sense. There is no reason to re-seal the same data just because one needs to spend new DataCap.

> fn ExtendClaimTerms(params: ExtendClaimTermsParams) -> ExtendClaimTermsResult

Reviewer: what's `maximum initial term` ? like 5 years?

Author: That is the value currently specified above, yes.

```

> A verified client can extend the term for a claim _beyond_ the initial maximum term by

Reviewer: a bit confused. above it says

> Called by allocation client to extend the maximum term of their allocations,  **without consuming DataCap**.

inconsistency maybe?

Reviewer: from the context below it seems like we meant
```suggestion
A verified client can extend the term for a claim _beyond_ the initial maximum term without
```

spending new DataCap.
The term maximum can be extended up to the network policy maximum beyond the current epoch.

> The client extending the claim need not be the one that made the original allocation.

Reviewer: above states

_maximum term for a claim may be increased **by the client which originally made the allocation**,_

but here says need _**not** be the one that made the original allocation._

Reviewer 3: These comments are all good points but I think they're getting resolved by deleting `ExtendClaimTerms` which will happen when the following TODO is addressed

Author: You're confused because there are two different mechanisms here:
1. The original client may increase term up to the maximum that could have been specified originally without spending any data cap
2. Any client can add an additional maximum term beyond the current epoch, by spending data cap

Reviewer: ah-ha!

but any subsequent penalty is calculated using its current (i.e. reduced) power.

### Storage miner actor
#### Differentiated sectors
A new field `HasVerifiedClaims: bool` is added to `SectorOnChainInfo`. 
This is set `false` for all existing sectors during migration (below).

> A _new_ sector takes `HasVerifiedClaims = true` if it claims one or more verified allocations.

Reviewer: ```suggestion
A _new_ sector takes `HasVerifiedClaims = true` if it claims one or more allocations.
```

Reviewer: suggested cuz "verified allocation" was not an terminology used up until this point. (not a big deal tho)

Its `DealWeight` and `VerifiedDealWeight` will be a simple sum of the size of those claims.
Its power is calculated in the simple way described above.

A _new_ sector takes `HasVerifiedClaims = false` if it claims no verified allocations.

A sector with `HasVerifiedClaims = false` and `VerifiedDealWeight = 0` 

> (i.e. a committed-capacity sector) takes `HasVerifiedClaims = true` if it claims 

Reviewer 3: nit: these don't really exist but it could also be non verified deal sector

one or more verified allocations by `ReplicaUpdate`.
Its new `DealWeight` and `VerifiedDealWeight` are calculated as a sum of the size of those claims.
Its power is re-calculated in the simple way described above.

> A sector with `HasVerifiedClaims = false` and `VerifiedDealWeight > 0` continues to use

Reviewer: will there ever exists a situation where a `HasVerifiedClaims` changes from _true_ to _false_, due to all claims in that sector reached term max gradually?

the existing QA-power calculation.
Such a sector cannot claim a verified allocation.
Such a sector may be extended, but, as today, this will result in its power being diluted
to "spread out" the verified reward over the sector's new commitment duration.

#### Term enforcement

> The built-in miner actor enforces the verified term and the continual storage of the data.

Reviewer: ```suggestion
The built-in miner actor enforces the claim term and the continual storage of the data.
```

but any subsequent penalty is calculated using its current (i.e. reduced) power.

### Storage miner actor
#### Differentiated sectors
A new field `SimpleQAPower: bool` is added to `SectorOnChainInfo`. 
During the migration activating this proposal, `SimpleQAPower` is set for each sector:
- `true` if the sector has `VerifiedDealWeight = 0` and `DealWeight = 0`,
- `false` otherwise

All sectors committed after this proposal is activated take `SimpleQAPower = true`.
The value of `SimpleQAPower` never changes for an individual sector.

A sector with `SimpleQAPower = true` has its deal weights calculated as simple sums of deal size,
and QA power calculated according to the simple method described above, with no spreading out effect.

A sector with `SimpleQAPower = false` continues to use the existing QA-power calculation.

> Such a sector cannot claim any allocations.

Reviewer 2: This is likely to result in increased expirations, with "old" sectors being replaced by new sectors that can take allocations. 

In addition, CC sectors can take advantage of this new way to package deals, while sectors with data have to wait for their deals to end. 

Could we have a mechanism to turn sectors with  `SimpleQAPower = false` into sectors with `SimpleQAPower = true`? For instance, if I have a 1-year sector with an old 1-year deal, there's no reason why I shouldn't be able to pack new deals into it, using the new logic and QAP calculation.

Author: It is impossible to change the data in a non-CC sector today _anyway_ (there is no "re-snap"), so there are no sectors to which this would apply. It will be easier to implement re-snapping after this change, though, but by the time we do this will be an even smaller issue.

There are [reasons](https://github.com/filecoin-project/FIPs/discussions/313#discussioncomment-3530843) not to do this. Perhaps for a narrow case of sectors where all the deals have already expired there would be fewer unwanted effects, but still this would cause a slight power loss for the SP, so is unlikely a major flow.

I agree that, for a capacity-constrained SP, a CC sector is more valuable than a sector with expired deals at the end of its life, due to the option of taking deals into it. But that's already the case today.

#### Sector extension
A sector's scheduled commitment term cannot be extended beyond the maximum term of any

> verified claim attributed to it (if `HasVerifiedClaims = true`).

Reviewer: ```suggestion claim attributed to it (if `HasVerifiedClaims = true`).
```

and maximum term to be some amount greater than the deal duration.

**Parameters**

> - `MarketDefaultMaximumAllocationTerm`: The built-in market calculates a term maximum of 90 days greater than a deal's specified duration.

Reviewer 2: Why 90 days?

Author: Why not? This value was chosen as a compromise between flexibility for SP deal packing, and reasonably requesting the client's nominated term. The discussion thread for this is https://github.com/filecoin-project/FIPs/discussions/313#discussioncomment-3442637

Reviewer 2: Trying to understand the reasoning. I don't have any arguments against the 90 days.

  Size: deal.PieceSize
  TermMinimum: deal.EndEpoch - deal.StartEpoch

>   TermMaximum: deal.EndEpoch - deal.StartEpoch + EPOCHS_IN_YEAR

Reviewer: ```suggestion
  TermMaximum: deal.EndEpoch - deal.StartEpoch + 90 * EPOCHS_IN_DAY
```

  Expiration: min(deal.StartEpoch, CURRENT_EPOCH + 60 * EPOCHS_IN_DAY)
  AllowRanges: false
  }
```

> The market will accept verified deals with a `StartEpoch` beyond the maximum allowed allocation `Expiration`.

Reviewer 3: So for miner integration: deal activation can pass and alloc claim can fail and in that case deal is still included, sector is still ProveCommitted but QA power is not added for the failed deal?

Reviewer 3: * in the case where prove commit happens after claim expiration but before deal start

Author: Yes

computes quality adjusted power for the sector according to the piece size and QA power multiplier.

### Migration
A network state migration is required to implement this proposal.

That migration must:
- Implement the new verified registry state schema for allocations and claims,

>   and set `NextAllocationId = 1`.

Reviewer 3: [I think this](https://github.com/filecoin-project/FIPs/pull/455/files#diff-da7f03d0785adde81d03c9123f3d0cf50d8633ca6ac08d826a561239ca7dc66cR138)

>   and set `NextAllocationId = 1`.

Reviewer 3: [And then used here](https://github.com/filecoin-project/FIPs/pull/455/files#diff-da7f03d0785adde81d03c9123f3d0cf50d8633ca6ac08d826a561239ca7dc66cR563)

>   and set `NextAllocationId = 1`.

Author: Answered above, 0 is reserved for a "no allocation" value

- Migrate the verified registry client balances map to DataCap token balances, 
  with appropriate unit conversion.
- Add a new field `HasVerifiedClaim = false` to each `SectorOnChainInfo`.
- Add the pending allocation mapping to market state, and claim ID field to deal states (all `0`)
- For each pending (not yet activated) deal with `Verified = true`, create an Allocation
  - The maximum term for such allocations is 540 days, in order to support any possible sector duration

>   - The expiration for such allocations is set to the maximum allowed value after the migration epoch,

Reviewer 3: By this you mean [60 days](https://github.com/filecoin-project/FIPs/pull/455/files#diff-da7f03d0785adde81d03c9123f3d0cf50d8633ca6ac08d826a561239ca7dc66cR124)?

Author: Yes
